File: src/main/java/com/pauljoda/modularsystems/power/tiles/TileBankBase.java
package com.pauljoda.modularsystems.power.tiles;

import com.pauljoda.modularsystems.core.providers.FuelProvider;
import com.pauljoda.modularsystems.core.tiles.DummyTile;
import com.teambr.bookshelf.api.waila.IWaila;
import com.teambr.bookshelf.helpers.GuiHelper;
import mcp.mobius.waila.api.IWailaConfigHandler;
import mcp.mobius.waila.api.IWailaDataAccessor;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;

import java.util.List;

/**
 * The base tile for all fuel banks
 */
public abstract class TileBankBase extends DummyTile implements FuelProvider, IWaila {

    //Current Priority
    protected int priority;

    public TileBankBase() {
        priority = 0;
    }

    /**
     * Used to scale the current power level
     * @param scale The scale to move to
     * @return A number from 0 - {@param scale} for current level
     */
    public abstract double getPowerLevelScaled(int scale);

    /*******************************************************************************************************************
     ******************************************* Tile Methods **********************************************************
     *******************************************************************************************************************/
    @Override
    public void readFromNBT (NBTTagCompound tags) {
        super.readFromNBT(tags);
        priority = tags.getInteger("priority");
    }

    @Override
    public void writeToNBT (NBTTagCompound tags) {
        super.writeToNBT(tags);
        tags.setInteger("priority", priority);
    }

    /*******************************************************************************************************************
     ***************************************** Fuel Provider Functions *************************************************
     *******************************************************************************************************************/

    @Override
    public int getPriority() {
        return priority;
    }

    public void setPriority(int value) {
        priority = value;
    }

    /*******************************************************************************************************************
     ************************************************ Waila ************************************************************
     *******************************************************************************************************************/

    @Override
    public NBTTagCompound returnNBTData(EntityPlayerMP player, TileEntity te, NBTTagCompound tag, World world, int x, int y, int z) {
        if (tag.hasKey("Energy")) {
            tag.removeTag("Energy");
            tag.removeTag("MaxStorage");
        }
        return tag;
    }

    @Override
    public void returnWailaTail(List<String> list) {
        list.add(GuiHelper.GuiColor.ORANGE + GuiHelper.GuiTextFormat.ITALICS.toString() + "Shift+Click to access GUI");
    }

    @Override
    public void returnWailaHead(List<String> list) {}

    @Override
    public void returnWailaBody(List<String> list) {}

    @Override
    public ItemStack returnWailaStack(IWailaDataAccessor accessor, IWailaConfigHandler config) { return null; }
}

File: src/main/scala/com/teambr/modularsystems/power/tiles/TileBankBase.scala
package com.teambr.modularsystems.power.tiles

import com.teambr.modularsystems.core.providers.FuelProvider
import com.teambr.modularsystems.core.common.tiles.TileProxy
import net.minecraft.nbt.NBTTagCompound

/**
 * This file was created for Modular-Systems
 *
 * Modular-Systems is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Dyonovan
 * @since August 07, 2015
 */
abstract class TileBankBase extends TileProxy with FuelProvider {

    protected[power] var priority = 0
    protected[power] var coolDown = 0

    /**
     * Used to scale the current power level
     * @param scale The scale to move to
     * @return A number from 0 - { @param scale} for current level
     */
    def getPowerLevelScaled(scale: Int): Double

    /******************************************************************************************************************
      ******************************************* Tile Methods ********************************************************
      *****************************************************************************************************************/
    override def readFromNBT(tag: NBTTagCompound) : Unit = {
        super.readFromNBT(tag)
        priority = tag.getInteger("priority")
    }

    override def writeToNBT(tag: NBTTagCompound) {
        super.writeToNBT(tag)
        tag.setInteger("priority", priority)
    }

    /*******************************************************************************************************************
      ***************************************** Fuel Provider Functions ************************************************
      ******************************************************************************************************************/
    override def getPriority: Int = {
        priority
    }
}
